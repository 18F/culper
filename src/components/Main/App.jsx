import React from 'react'
import { withRouter } from 'react-router'
import { connect } from 'react-redux'
import { i18n } from '../../config'
import {
  SectionTitle,
  ProgressBar,
  ScoreCard,
  Navigation,
  NavigationToggle
} from '..'
import { Form } from '../../views'
import { Introduction, Show } from '../Form'
import Logout from '../Navigation/Logout'
import StickyHeader from '../Sticky/StickyHeader'
import formTypes from './../../config/formTypes'

/*
           1/6-ish                                 2/3-ish                               1/6-ish
  ------------------------------------------------------------------------------------------------
  |                       |
  | Title                 | <Page title />
  | <Logout button />     |
  |                       |
  ------------------------------------------------------------------------------------------------
  |                       |
  | <Score />             | <View />
  | <Nav />               |  - This will be where the main content lies
  |                       |
  |                       |
  |                       |
  |                       |
  |                       |
  |                       |
  ------------------------------------------------------------------------------------------------
  |
  | <Footer />
  |
  ------------------------------------------------------------------------------------------------
*/

class App extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      instructions: false
    }
    this.showInstructions = this.showInstructions.bind(this)
    this.dismissInstructions = this.dismissInstructions.bind(this)

    // workaround for not having React.createRef(), introduced in React 16.3
    // https://reactjs.org/docs/refs-and-the-dom.html#dont-overuse-refs
    this.sectionFocusEl = null
    this.setSectionFocusEl = el => {
      this.sectionFocusEl = el
    }
  }

  componentDidUpdate(prevProps) {
    // for keyboard navigation accessbility, focus on the main content area after a new section is navigated to
    if (this.props.location.pathname !== prevProps.location.pathname) {
      this.sectionFocusEl.focus()
    }
  }

  showInstructions(event) {
    this.setState({ instructions: true })
  }

  dismissInstructions() {
    this.setState({ instructions: false })
  }

  designClass() {
    if (
      this.props.location &&
      this.props.location.query &&
      this.props.location.query.design
    ) {
      return 'design'
    }

    if (
      window &&
      window.location &&
      window.location.search &&
      window.location.search.indexOf('design') !== -1
    ) {
      return 'design'
    }

    return ''
  }

  render() {
    const klassApp = `${this.designClass()} ${
      this.props.settings.modalOpen ? 'modal-open' : ''
    }`.trim()
    const mobileNavigation = this.props.settings.mobileNavigation || false
    const klassTitle = 'eapp-structure-right eapp-title'
    const klassHeading = `eapp-structure-wrap eapp-header ${
      mobileNavigation ? 'mobile-navigation' : ''
    }`.trim()
    const klassMain = `eapp-structure-wrap eapp-main ${
      mobileNavigation ? 'mobile-navigation' : ''
    }`.trim()
    const klassNavigation = `eapp-structure-left eapp-navigation ${
      mobileNavigation
        ? 'mobile-visible tablet-visible desktop-visible'
        : 'mobile-hidden'
    }`
    const klassCore = `eapp-structure-right eapp-core ${
      mobileNavigation ? 'mobile-hidden' : 'visible'
    }`

    return (
      <div className={klassApp}>
        <Introduction
          forceOpen={this.state.instructions}
          onDismiss={this.dismissInstructions}
          dispatch={this.props.dispatch}
        />
        <div id="scrollTo" />
        <a className="usa-skipnav" href="#main-content">
          {i18n.t('app.skip')}
        </a>
        <StickyHeader stickyClass="sticky-header">
          <div className="header-container">
            <div className="header">
              <header className="usa-header usa-header-basic" role="banner">
                <div className="usa-banner mobile-hidden">
                  <div className="usa-accordion">
                    <div className="usa-banner-header">
                      <div className="usa-grid usa-banner-inner">
                        <img
                          src="/img/favicons/favicon-57.png"
                          alt="U.S. flag"
                        />
                        <p>{i18n.t('app.banner.title')}</p>
                        <button
                          className="usa-accordion-button usa-banner-button"
                          aria-expanded="false"
                          aria-controls="gov-banner">
                          <span className="usa-banner-button-text">
                            {i18n.t('app.banner.button')}
                          </span>
                        </button>
                      </div>
                    </div>
                    <div
                      className="usa-banner-content usa-grid usa-accordion-content"
                      id="gov-banner">
                      <div className="usa-banner-guidance-gov usa-width-one-half">
                        <img
                          className="usa-banner-icon usa-media_block-img"
                          src="/img/icon-dot-gov.svg"
                          alt="Dot gov"
                        />
                        <div className="usa-media_block-body">
                          <p>
                            <strong>{i18n.t('app.banner.witty')}</strong>
                            <br />
                            {i18n.t('app.banner.extension')}
                          </p>
                        </div>
                      </div>
                      <div className="usa-banner-guidance-ssl usa-width-one-half">
                        <img
                          className="usa-banner-icon usa-media_block-img"
                          src="/img/icon-https.svg"
                          alt="SSL"
                        />
                        <div className="usa-media_block-body">
                          <p>{i18n.t('app.banner.ssl')}</p>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div className={klassHeading}>
                  <div className="eapp-structure-row">
                    <div className="eapp-structure-left eapp-logo" id="logo">
                      <NavigationToggle />
                      <div className="eapp-logo-container">
                        <img
                          className="eapp-logo-icon"
                          src="/img/nbis-seal-small.png"
                          alt="National Background Investigation Services"
                        />
                        <span className="eapp-logo-text">
                          {formTypes[this.props.formType]}
                        </span>
                      </div>
                    </div>
                    <div className={klassTitle}>
                      <div className="eapp-logout">
                        <button
                          onClick={this.showInstructions}
                          className="instructions mobile-hidden">
                          {i18n.t('app.instructions')}
                        </button>
                        <Show when={this.props.authenticated}>
                          <Logout />
                        </Show>
                      </div>
                      <SectionTitle hidden={mobileNavigation} />
                    </div>
                  </div>
                </div>
              </header>
              <div id="scrollToProgress" />
              <div className="usa-overlay" />
              <ProgressBar />
            </div>
          </div>
        </StickyHeader>
        <main className={klassMain}>
          <div className="eapp-structure-row">
            <div className={klassNavigation}>
              <ScoreCard />
              <Navigation />
              <button
                onClick={this.showInstructions}
                className="instructions mobile-visible">
                <span>{i18n.t('app.instructions')}</span>
              </button>
              &nbsp;
            </div>
            <a
              href="javascript:;;;"
              className="eapp-section-focus"
              title="Main content. Please press TAB to go to the next question"
              ref={this.setSectionFocusEl}
            />
            <div id="main-content" className={klassCore}>
              <Form {...this.props} />
              &nbsp;
            </div>
          </div>
        </main>
      </div>
    )
  }
}

/**
 * Maps the relevant subtree state from the applications state tree.
 * In this case, we pull the authentication sub-state. This is mapped
 * to the authentication reducer. When actions are dispatched, this
 * method is executed which causes a re-render.
 *
 */
function mapStateToProps(state) {
  const auth = state.authentication
  const app = state.application || {}
  const settings = app.Settings || { mobileNavigation: false, modalOpen: false }

  return {
    settings: settings,
    authenticated: auth.authenticated,
    formType: auth.formType
  }
}

// Wraps the the App component with connect() which adds the dispatch()
// function to the props property for this component
export default withRouter(connect(mapStateToProps)(App))
