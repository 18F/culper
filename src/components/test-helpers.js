import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'

export const testSnapshot = (jsx, minElements = 3) => {
  // Sanity check to make sure the component is being rendered with the full subtree. Using Enzyme as a workaround for not having the find*() methods available in React 16

  const middlewares = [thunk]
  const mockStore = configureMockStore(middlewares)
  const store = mockStore()

  const enzymeComponent = mount(<Provider store={store}>{jsx}</Provider>)

  expect(enzymeComponent.find('div').length).toBeGreaterThan(minElements)

  const component = renderer.create(<Provider store={store}>{jsx}</Provider>)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
}
