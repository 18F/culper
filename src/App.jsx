import React from 'react'
import { i18n } from './config'
import { SectionTitle, ProgressBar, Sticky, ScoreCard, Navigation, NavigationToggle } from './components'
import { Introduction } from './components/Form'
import StickyHeader from './components/Sticky/StickyHeader'
import { connect } from 'react-redux'
import { env } from './config'
import { api } from './services/api'
import { logout } from './actions/AuthActions'

/*
           1/6-ish                                 2/3-ish                               1/6-ish
  ------------------------------------------------------------------------------------------------
  |                       |
  | Title                 | <Page title />
  | <Logout button />     |
  |                       |
  ------------------------------------------------------------------------------------------------
  |                       |
  | <Score />             | <View />
  | <Nav />               |  - This will be where the main content lies
  |                       |
  |                       |
  |                       |
  |                       |
  |                       |
  |                       |
  ------------------------------------------------------------------------------------------------
  |
  | <Footer />
  |
  ------------------------------------------------------------------------------------------------
*/

class App extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      instructions: false
    }
    this.logout = this.logout.bind(this)
    this.showInstructions = this.showInstructions.bind(this)
    this.dismissInstructions = this.dismissInstructions.bind(this)
  }

  logout () {
    this.props.dispatch(logout())
    window.location = window.location.pathname
  }

  showInstructions (event) {
    this.setState({ instructions: true })
  }

  dismissInstructions () {
    this.setState({ instructions: false })
  }

  designClass () {
    if (this.props.location && this.props.location.query && this.props.location.query.design) {
      return 'design'
    }

    if (window && window.location && window.location.search && window.location.search.indexOf('design') !== -1) {
      return 'design'
    }

    return ''
  }

  render () {
    validWebToken()
      .then(token => {
        if (!token) {
          this.props.dispatch(logout())
        }
      })
      .catch(() => {
        this.props.dispatch(logout())
      })

    const logoutButton = this.props.authenticated || this.props.twofactor
        ? (<a href="#" onClick={this.logout} className="logout">{i18n.t('app.logout')}</a>)
        : null
    const klassApp = `${this.designClass()} ${this.props.settings.modalOpen ? 'modal-open' : ''}`.trim()
    const klassTitle = `eapp-structure-right eapp-title ${this.props.settings.mobileNavigation ? 'mobile-hidden' : 'visible'}`.trim()
    const klassNavigation = `eapp-structure-left eapp-navigation ${this.props.settings.mobileNavigation ? 'mobile-visible' : 'mobile-hidden'}`.trim()
    const klassCore = `eapp-structure-right eapp-core ${this.props.settings.mobileNavigation ? 'mobile-hidden' : 'visible'}`.trim()

    return (
      <div className={klassApp}>
        <Introduction forceOpen={this.state.instructions}
                      onDismiss={this.dismissInstructions}
                      dispatch={this.props.dispatch} />
        <div id="scrollTo"></div>
        <a className="usa-skipnav" href="#main-content">{i18n.t('app.skip')}</a>
        <StickyHeader stickyClass="sticky-header">
          <div className="header-container">
            <div className="header">
              <header className="usa-header usa-header-basic" role="banner">
                <div className="usa-banner mobile-hidden">
                  <div className="usa-accordion">
                    <header className="usa-banner-header">
                      <div className="usa-grid usa-banner-inner">
                        <img src="/img/favicons/favicon-57.png" alt="U.S. flag" />
                        <p>{i18n.t('app.banner.title')}</p>
                        <button className="usa-accordion-button usa-banner-button"
                          aria-expanded="false" aria-controls="gov-banner">
                          <span className="usa-banner-button-text">{i18n.t('app.banner.button')}</span>
                        </button>
                      </div>
                    </header>
                    <div className="usa-banner-content usa-grid usa-accordion-content" id="gov-banner">
                      <div className="usa-banner-guidance-gov usa-width-one-half">
                        <img className="usa-banner-icon usa-media_block-img" src="/img/icon-dot-gov.svg" alt="Dot gov" />
                        <div className="usa-media_block-body">
                          <p>
                            <strong>{i18n.t('app.banner.witty')}</strong>
                            <br />
                            {i18n.t('app.banner.extension')}
                          </p>
                        </div>
                      </div>
                      <div className="usa-banner-guidance-ssl usa-width-one-half">
                        <img className="usa-banner-icon usa-media_block-img" src="/img/icon-https.svg" alt="SSL" />
                        <div className="usa-media_block-body">
                          <p>{i18n.t('app.banner.ssl')}</p>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div className="eapp-structure-wrap eapp-header">
                  <div className="eapp-structure-row">
                    <div className="eapp-structure-left eapp-logo" id="logo">
                      <img className="eapp-logo-icon" src="/img/US-OfficeOfPersonnelManagement-Seal.svg" alt="Office of Personnet Management" />
                      <span className="eapp-logo-text">SF86</span>
                      <NavigationToggle />
                    </div>
                    <div className={klassTitle}>
                      <div className="eapp-logout mobile-hidden">
                        <button onClick={this.showInstructions} className="instructions">{i18n.t('app.instructions')}</button>
                        {logoutButton}
                      </div>
                      <SectionTitle />
                    </div>
                  </div>
                </div>
              </header>
              <div id="scrollToProgress"></div>
              <div className="usa-overlay"></div>
              <ProgressBar />
            </div>
          </div>
        </StickyHeader>
        <main className="eapp-structure-wrap">
          <div className="eapp-structure-row">
            <div className={klassNavigation}>
              <Sticky options={{tolerance: 400, ignoreWindowComparison: true}}>
                <ScoreCard />
                <Navigation />
              </Sticky>
              &nbsp;
            </div>
            <div id="main-content" className={klassCore}>
              <Sticky container=".eapp-navigation .sticky > .contents"
                      content=".eapp-core .sticky > .contents"
                      options={{tolerance: 400, ignoreWindowComparison: true}}>
                {this.props.children}
              </Sticky>
              &nbsp;
            </div>
          </div>
        </main>
      </div>
    )
  }
}

/**
 * Maps the relevant subtree state from the applications state tree.
 * In this case, we pull the authentication sub-state. This is mapped
 * to the authentication reducer. When actions are dispatched, this
 * method is executed which causes a re-render.
 *
 */
function mapStateToProps (state) {
  const auth = state.authentication
  const app = state.application || {}
  const settings = app.Settings || { mobileNavigation: false, modalOpen: false }

  return {
    settings: settings,
    authenticated: auth.authenticated,
    twofactor: auth.twofactor
  }
}

// Wraps the the App component with connect() which adds the dispatch()
// function to the props property for this component
export default connect(mapStateToProps)(App)

const validWebToken = () => {
  return new Promise((resolve, reject) => {
    if (env.IsTest()) {
      console.log('Skip refreshing web tokens')
      resolve(api.getToken())
    }

    const token = api.getToken()
    if (!token) {
      reject()
      return
    }

    api.refresh()
      .then(r => {
        api.setToken(r.data)
        resolve(r.data)
      })
      .catch(() => {
        api.setToken('')
        reject()
      })
  })
}
