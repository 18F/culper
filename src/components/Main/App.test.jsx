import React from 'react'
import { MemoryRouter } from 'react-router'
import App from './App'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'
import { testSnapshot } from '../test-helpers'

// give a fake GUID so the field IDs don't differ between snapshots
// https://github.com/facebook/jest/issues/936#issuecomment-404246102
jest.mock('../Form/ValidationElement/helpers', () =>
  Object.assign(require.requireActual('../Form/ValidationElement/helpers'), {
    newGuid: jest.fn().mockReturnValue('MOCK-GUID')
  })
)

test('Renders homepage', () => {
  // Setup
  const middlewares = [thunk]
  const mockStore = configureMockStore(middlewares)
  const store = mockStore({
    authentication: [],
    application: {
      Settings: {
        mobileNavigation: false
      },
      Navigation: {
        sections: [],
        totalSections: 10,
        completedSectionsTotal: 0
      },
      Completed: {}
    }
  })

  testSnapshot(
    <MemoryRouter>
      <Provider store={store}>
        <App />
      </Provider>
    </MemoryRouter>
  )
})
