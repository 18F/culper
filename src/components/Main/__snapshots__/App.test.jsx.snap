// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders homepage 1`] = `
<div
  className=""
>
  <div
    id="scrollTo"
  />
  <a
    className="usa-skipnav"
    href="#main-content"
  >
    Skip to form questions
  </a>
  <div
    className="sticky-header "
  >
    <div
      className="header-container"
    >
      <div
        className="header"
      >
        <header
          className="usa-header usa-header-basic"
          role="banner"
        >
          <div
            className="usa-banner mobile-hidden"
          >
            <div
              className="usa-accordion"
            >
              <div
                className="usa-banner-header"
              >
                <div
                  className="usa-grid usa-banner-inner"
                >
                  <img
                    alt="U.S. flag"
                    src="/img/favicons/favicon-57.png"
                  />
                  <p>
                    An official website of the United States government
                  </p>
                  <button
                    aria-controls="gov-banner"
                    aria-expanded="false"
                    className="usa-accordion-button usa-banner-button"
                  >
                    <span
                      className="usa-banner-button-text"
                    >
                      Here's how you know
                    </span>
                  </button>
                </div>
              </div>
              <div
                className="usa-banner-content usa-grid usa-accordion-content"
                id="gov-banner"
              >
                <div
                  className="usa-banner-guidance-gov usa-width-one-half"
                >
                  <img
                    alt="Dot gov"
                    className="usa-banner-icon usa-media_block-img"
                    src="/img/icon-dot-gov.svg"
                  />
                  <div
                    className="usa-media_block-body"
                  >
                    <p>
                      <strong>
                        The .gov means it's official.
                      </strong>
                      <br />
                      Federal government websites always use a .gov or .mil domain. Before sharing sensitive information online, make sure you're on a .gov or .mil site by inspecting your browser's address (or "location") bar.
                    </p>
                  </div>
                </div>
                <div
                  className="usa-banner-guidance-ssl usa-width-one-half"
                >
                  <img
                    alt="SSL"
                    className="usa-banner-icon usa-media_block-img"
                    src="/img/icon-https.svg"
                  />
                  <div
                    className="usa-media_block-body"
                  >
                    <p>
                      This site is also protected by an SSL (Secure Sockets Layer) certificate that's been signed by the U.S. government. The &lt;strong&gt;https://&lt;/strong&gt; means all transmitted data is encrypted  — in other words, any information or browsing history that you provide is transmitted securely.
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="eapp-structure-wrap eapp-header"
          >
            <div
              className="eapp-structure-row"
            >
              <div
                className="eapp-structure-left eapp-logo"
                id="logo"
              >
                <div
                  className="text-center"
                >
                  <img
                    alt="Office of Personnet Management"
                    className="eapp-logo-icon"
                    src="/img/US-OfficeOfPersonnelManagement-Seal.svg"
                  />
                  <span
                    className="eapp-logo-text"
                  >
                    SF86
                  </span>
                </div>
              </div>
              <div
                className="eapp-structure-right eapp-title"
              >
                <div
                  className="eapp-logout"
                >
                  <button
                    className="instructions mobile-hidden"
                    onClick={[Function]}
                  >
                    Instructions
                  </button>
                </div>
                <h1
                  className="title"
                />
              </div>
            </div>
          </div>
        </header>
        <div
          id="scrollToProgress"
        />
        <div
          className="usa-overlay"
        />
      </div>
    </div>
  </div>
  <main
    className="eapp-structure-wrap eapp-main"
  >
    <div
      className="eapp-structure-row"
    >
      <div
        className="eapp-structure-left eapp-navigation mobile-hidden"
      >
        <button
          className="instructions mobile-visible"
          onClick={[Function]}
        >
          <span>
            Instructions
          </span>
        </button>
         
      </div>
      <a
        className="eapp-section-focus"
        href="javascript:;;;"
        title="Main content. Please press TAB to go to the next question"
      />
      <div
        className="eapp-structure-right eapp-core visible"
        id="main-content"
      >
         
      </div>
    </div>
  </main>
</div>
`;
